lib/components/user/user-account-screen.js

```diff
@@ -0,0 +1,437 @@
+        persistence.otp_middleware,
+        'auth0UserId',
+        auth.user.sub,
+        auth.accessToken
+      )
+
+      /**
+       * Beware! On AWS, for a nonexistent user, the call above will return, for example:
+       * {
+       *    status: 'success',
+       *    data: {
+       *      "result": "ERR",
+       *      "message": "No user with auth0UserId=000000 found.",
+       *      "code": 404,
+       *      "detail": null
+       *    }
+       * }
+       *
+       * On direct middleware interface, for a nonexistent user, the call above will return:
+       * {
+       *    status: 'error',
+       *    message: 'Error get-ing user...'
+       * }
+       * TODO: Improve AWS response.
+       */
+      const isNewAccount = result.status === 'error' || (result.data && result.data.result === 'ERR')
+
+      if (!isNewAccount && skipIfExistingUser) {
+        // Return to originalUrl and prevent the URL from showing in browsing history
+        window.location.replace(window.location.href.replace('#/account/welcome', `#${originalUrl}`))
+      } else {
+        // TODO: Set application state (so that user data is fetched once only... until logout happens).
+        const userData = isNewAccount ? getStateForNewUser(auth.user) : result.data
+        if (userData.savedLocations === null) userData.savedLocations = []
+
+        this.setState({
+          isNewAccount,
+          userData
+        })
+      }
+    } catch (error) {
+      // TODO: improve error handling.
+      alert(`An error was encountered:\n${error}`)
+    }
+  }
+
+  _updateUserState = newUserData => {
+    const { userData } = this.state
+    this.setState({
+      userData: {
+        ...userData,
+        ...newUserData
+      }
+    })
+  }
+
+  _updateUserPrefs = async () => {
+    const { auth, persistence } = this.props
+
+    if (persistence && persistence.otp_middleware) {
+      const { accessToken } = auth
+      const { isNewAccount, userData } = this.state
+
+      // TODO: Change state of Save button.
+
+      let result
+      if (isNewAccount) {
+        result = await addUser(persistence.otp_middleware, accessToken, userData)
+      } else {
+        result = await updateUser(persistence.otp_middleware, accessToken, userData)
+      }
+
+      // TODO: improve this.
+      if (result.status === 'success') {
+        alert('Your preferences have been saved.')
+      } else {
+        alert(`An error was encountered:\n${result.message}`)
+      }
+    }
+  }
+
+  _handleConsentChange = e => {
+    this._updateUserState({ hasConsentedToTerms: e.target.checked })
+  }
+
+  _handleExit = () => {
+    const { originalUrl } = this.props
+    this.props.routeTo(originalUrl)
+  }
+
+  _handleExitAndSave = async () => {
+    await this._updateUserPrefs()
+    this._handleExit()
+  }
+
+  _handleLocationsChange = newLocations => {
+    this._updateUserState({ savedLocations: newLocations })
+  }
+
+  _handleNotificationChannelChange = e => {
```

Comment:
All these special handlers that only call updateUserState could be contained within the individual components for each pane.

Reply:
Good idea